### DISCUSSION TITLE
Secrets Handling in AL

### DISCUSSION THREAD

**Discussion by StefanMaron** [microsoft/alguidelines#251] [BC Best Practices]
I am quite surprised this was not brought up yet, so let me start this :)

What would be the best practice way of handling secret values in an App?

Before we got the `IsolatedStorage` they where just saved in a setup table with `Masked=true` which is of course not ideal.

I started to implement them with Isolated storage now, with a Codeunit that handles all the getting and setting for me.
I also created a Page to let the user enter the Secret and I show on the Setup page if the Secret is set or not.

I would like to know if I am missing something obvious or if there is still room for improvement.

Once the idea is complete I will happily summarize this in a Docs page here :)

**Comment by NKarolak** [comment]
> I would like to know if I am missing something obvious.

Not to my knowledge, as this is exactly how I have implemented it as well.

**Comment by ChrisBlankDe** [comment]
> Codeunit that handles all the getting and setting 

centralized codeunits should not provide a getter method. this would be an security issue.

**Reply by StefanMaron** [reply]
Can you elaborate on this?

**Reply by ChrisBlankDe** [reply]
you have motivated me to finish my blog post, which has been hanging around for almost a year: https://www.chris-blank.de/2024/10/08/isolated-storage-is-no-guarantee-of-security/

Long story short: internal procedures can be called by other apps

**Reply by StefanMaron** [reply]
But what I get from this is, that firstly we need to make sure to remove the internals visible to from the app.json in the pipeline before publishing to github. Its mainly used to expose procedures to test apps anyways.

Secondly, I think this would prevent any of the security concerns, right?

https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/navapp/navapp-getcallermoduleinfo-method

Good blog btw :)

**Reply by ChrisBlankDe** [reply]
do not remove `internalsvisibleto`.  just move the GetPassword() Proc to the codeunit wher its been used and change the access modifier to local.
that would resolve my security concern.


GetCallerModuleInfo would just return the faked app info. this function cant be used for security topics.

**Reply by StefanMaron** [reply]
Of course it would return that, but I can verify that the caller is never another app that my own app. 

Having the getter functions makes it way easier to handle the secrets tho

**Reply by ChrisBlankDe** [reply]
ah ok. didn't get your intention at first.
checking GetCallerModuleInfo AppId against GetCurrentModuleInfo AppId would be secure but imho kind of strange.

i dont see any problem in moving the code.
not to beat around the bush, here is the code we use most of the time:
```al
 [NonDebuggable]
 local procedure GetPassword() sText: SecretText;
 var
     StorageKeyTok: 'AnyGuid', Locked = true; // actually not a label but the retrun value of a centralized global function
 begin
     if IsolatedStorage.Contains(StorageKeyTok, DataScope::Module) then
         IsolatedStorage.Get(StorageKeyTok, DataScope::Module, sText);
 end;
```
imho this can be duplicated and does not have to be centralized

**Reply by StefanMaron** [reply]
If it is just that, agree. But sometimes more logic can be involved reading a value

**Reply by StefanMaron** [reply]
Just to give an example

![image](https://github.com/user-attachments/assets/5360bd57-02ad-4176-a85c-09cf9286e387)

**Comment by ChrisBlankDe** [comment]
It should also be mentioned to always use [secrettext](https://learn.microsoft.com/de-de/dynamics365/business-central/dev-itpro/developer/devenv-secret-text) as data type if possible. this is available since 2023w2 and got usable by 2024w1.

**Comment by curateideas** [comment]
What is a secure way to transfer screts between systems (i. e. from NAV to BC if it is not a direct migration)?

**Comment by ChrisBlankDe** [comment]
Looking at the screenshot from you (https://github.com/microsoft/alguidelines/discussions/251#discussioncomment-10879450) im thinking about that all secrets that belong together should also be stored together.
in the given example its a access token and its temporal validity.

i would sewrialize them to a json object and store it as one secret.
Benefit: less sql transaction. remeber every contains, set, get is a database transaction with no control about locking and readuncommited behaviour.

**Reply by StefanMaron** [reply]
I think that brings us into an Area of personal preference already. I like to keep my Database at least a bit normalized. I would guess with caching and optimizations the extra SQL queries are not that impact-full. 

But the point is very valid. If getter functions are used, `GetCallerInfo` should be used to make sure the secret can not be read from outside the app. I also created a feature request on Al-Go to be able to remove the `InternalsVisibleTo` from the appjson before publish.

I agree that this is still not the ideal solution but I saw that AlOps also offers this to at least decrease the likelihood of internals being exposed by accident.
https://github.com/microsoft/AL-Go/issues/1248
